### PR TITLE
Consolidate losses and evals

### PR DESCRIPTION
**Pull request by danbraunai**

### CODE REVIEW

File: spd/registry.py
Comment (danbraunai) [inline spd/registry.py:1]: Unrelated, just cleaning up

File: spd/run_spd.py [hunk 1]
```diff
@@ -37,17 +35,14 @@
...
     get_lr_with_warmup,
 )
 from spd.utils.module_utils import replace_std_values_in_layernorm
 from spd.utils.run_utils import save_file
 
 
-def local_log(
-    data: Mapping[str, float | Image.Image | wandb.plot.CustomChart], step: int, out_dir: Path
-) -> None:
+def local_log(data: dict[str, Any], step: int, out_dir: Path) -> None:
```
Comment (danbraunai) [inline spd/run_spd.py:45]: The data values will raise errors if they're not json serializable or not handled specifically in the function.

**Comment by oli-clive-griffin** [comment]
Quickly:
> Added https://github.com/goodfire-ai/spd/issues/176 to convert AliveComponentsTracker to a Metric

This is awesome.

File: spd/configs.py [hunk 2]
```diff
@@ -22,55 +19,182 @@
 from spd.experiments.tms.configs import TMSTaskConfig
 from spd.log import logger
 from spd.models.components import GateType
+from spd.models.sigmoids import SigmoidTypes
 from spd.spd_types import ModelPath, Probability
 
+#### Train Metric Configs ####
+TrainMetricClassname = Literal[
```
Comment (oli-clive-griffin) [inline spd/configs.py:26]: this is unused AFAICT

File: spd/configs.py [hunk 3]
```diff
@@ -22,55 +19,182 @@
...
+    "StochasticReconSubsetCEAndKL",
+]
+
+
+class EvalMetricConfig(BaseModel):
+    model_config: ClassVar[ConfigDict] = ConfigDict(extra="forbid", frozen=True)
+
+    slow: bool = Field(
+        default=False,
+        description=("If True, only run this metric once every slow_eval_freq steps."),
```
Comment (oli-clive-griffin) [inline spd/configs.py:102]: Do you want this to be configurable? My assumption is no, in which case could we just make it a static `ClassVar`?

Comment (oli-clive-griffin) [reply spd/configs.py:102]: (goes for all other slow metrics)

Comment (oli-clive-griffin) [reply spd/configs.py:102]: actually - Why not just put it on the metric computation class, then skip by `metric.slow` (as a static classvar again)

File: spd/configs.py [hunk 4]
```diff
@@ -326,22 +410,63 @@ def microbatch_size(self) -> PositiveInt:
...
+        loss_metric_configs. This is done by DEPRECATED_CONFIG_KEYS. This means that old runs which
+        have these keys will not be able to be used for e.g. finetuning, but the raw model will be
+        the same.
+        """
+
+        # We don't bother doing the complex mapping from an old structure with the ``eval_metrics``
+        # key to the new structure with the ``eval_metric_configs`` key.
+        config_dict.pop("eval_metrics", None)
+
+        # Remove SubsetReconstructionLoss if it appears
```
Comment (oli-clive-griffin) [inline spd/configs.py:446]: Why do we specifically need to support this one? seems like surely there shouldn't already be deprecated `Metric`s before this PR is even merged. Am I missing something?

File: spd/configs.py [hunk 5]
```diff
@@ -357,15 +482,6 @@ def handle_deprecated_config_keys(cls, config_dict: dict[str, Any]) -> dict[str,
...
     def validate_model(self) -> Self:
-        # If any of the coeffs are 0, raise a warning
-        msg = "is 0, you may wish to instead set it to null to avoid calculating the loss"
-        if self.ci_recon_coeff == 0:
-            logger.warning(f"recon_coeff {msg}")
-        if self.importance_minimality_coeff == 0:
-            logger.warning(f"importance_minimality_coeff {msg}")
-        if self.faithfulness_coeff == 0:
-            logger.warning(f"faithfulness_coeff {msg}")
-
```
Comment (oli-clive-griffin) [inline spd/configs.py:368]: 😍

File: spd/losses.py [hunk 6]
```diff
@@ -4,295 +4,128 @@
...
-    total_loss += config.importance_minimality_coeff * importance_minimality_loss
-    loss_terms["importance_minimality"] = importance_minimality_loss.item()
-
-    loss_terms["total"] = total_loss.item()
-
-    return total_loss, loss_terms
+    total = torch.tensor(0.0, device=batch.device)
+    terms: dict[str, float] = {}
+
+    for cfg in loss_metric_configs:
```
Comment (oli-clive-griffin) [inline spd/losses.py:52]: I have an idea here, noting as reminder for myself

File: spd/models/component_model.py [hunk 7]
```diff
@@ -107,12 +107,11 @@ def __init__(
         self.target_model = target_model
         self.C = C
         self.pretrained_model_output_attr = pretrained_model_output_attr
-
-        module_paths = get_target_module_paths(target_model, target_module_patterns)
+        self.module_paths = get_target_module_paths(target_model, target_module_patterns)
```
Comment (oli-clive-griffin) [inline spd/models/component_model.py:110]: nice, does this replace `list(component_model.components.keys())` throughout the codebase?

File: spd/configs.py (hunk 3 above)
Comment (danbraunai) [reply spd/configs.py:102]: I did have in mind for it to be configurable. But now that you say it, I suppose people can just hardcode class variable if they want to change it for an eval they're particularly interested in sometimes, that's not a big deal.

File: spd/models/component_model.py (hunk 7 above)
Comment (danbraunai) [reply spd/models/component_model.py:110]: Oh yeah that's nice. Just pushed a change with this.

File: spd/configs.py (hunk 2 above)
Comment (danbraunai) [reply spd/configs.py]: Removed.

File: spd/configs.py (hunk 3 above)
Comment (danbraunai) [reply spd/configs.py]: Now it's only in classvars.

File: spd/metrics/base.py [hunk 8]
```diff
@@ -1,194 +1,36 @@
...
-    3. Implement `compute()` to calculate final metric values
-
-    For distributed evaluation:
-    - Call `sync_dist()` before `compute()` to synchronize states across ranks
-    - This gives you explicit control over when synchronization happens
-    """
+    """Interface for metrics that can be used in training and evaluation."""
 
-    is_differentiable: bool | None = None
+    slow: bool = False
```
[the remark this answers is not in the document]
Comment (oli-clive-griffin) [reply spd/metrics/base.py:17]: (I think that's how you do it)

File: spd/metrics/README.md
Comment (oli-clive-griffin) [inline spd/metrics/README.md:1]: just noting that the modifications my branch had to this file were completely vibe-coded and not checked

File: spd/scripts/compare_models/compare_models.py [hunk 9]
```diff
@@ -253,7 +253,7 @@ def compute_activation_densities(
                 batch = extract_batch_data(next(eval_iterator))
                 batch = batch.to(self.device)
                 _, pre_weight_acts = model(
-                    batch, mode="pre_forward_cache", module_names=list(model.components.keys())
+                    batch, mode="pre_forward_cache", module_names=model.module_paths
```
Comment (oli-clive-griffin) [inline spd/scripts/compare_models/compare_models.py:256]: thoughts on making module_names=model.module_paths the default behaviour?

```
def forward(
   ...
   module_names: list[str] | None
   ...
)
   module_names = module_names or self.module_paths
```

**Review by oli-clive-griffin — APPROVED** [review]
Nice. love how this has turned out, feel free to merge after addressing

**Comment by oli-clive-griffin** [comment]
> I don't love the current structure of our metric configs:
> 1. Most of them just have a classname and nothing else
> 2. The classname itself corresponds to the name of the actual Metric class, but it could be anything since we just use it as a discriminated union.
> But I can't immediately think of a cleaner option. Note that even though they don't have extra params in the config, some are initialized with different arguments.

I honestly think this is completely fine.

File: spd/metrics/README.md
[the remark this answers is not in the document]
Comment (danbraunai) [reply spd/metrics/README.md:1]: Agreed, I'm just deleting this file. There's a lot of examples for people to read, and everything is explicit enough now.

File: spd/utils/distributed_utils.py [hunk 10]
```diff
@@ -98,8 +98,8 @@ def init_distributed(backend: Literal["nccl", "gloo"] | None = None) -> Distribu
             device_id=local_device,
         )
 
-    # Set the default cuda device for this process
-    if torch.cuda.is_available():
+    # Set the default cuda device for this process (only for NCCL backend)
+    if backend == "nccl" and torch.cuda.is_available():
```
[the remark this answers is not in the document]
Comment (danbraunai) [reply spd/utils/distributed_utils.py:102]: I've removed this. But, the reason it was there:

For some distributed tests we use gloo and two processes (mpirun -np 2 ...). If we run these tests on a machine that has a gpu available, then, unless we also guard this by `backend == "nccl"`, it will set the cuda device with some non-zero rank. This was an issue because, when running tests on a machine with one gpu but with multiple processes, it would try to use different gpu device ids.

I've fixed this by just forcing the device to be "cpu" in distributed_utils.get_device() when backend == gloo so it never tries to use a gpu.

**Review by oli-clive-griffin — APPROVED** [review]
Beautiful 👨‍🍳🤌
